pallets/amm/src/lib.rs

```diff
@@ -273,77 +289,73 @@ pub mod pallet {
 
             let (_, base_asset, quote_asset) = Self::get_upper_currency(pool.0, pool.1);
 
-            ensure!(
-                Pools::<T, I>::contains_key(base_asset, quote_asset),
-                Error::<T, I>::PoolDoesNotExist
-            );
-
-            let pool_liquidity_amount: PoolLiquidityAmount = Self::pools(base_asset, quote_asset);
-
-            ensure!(
-                pool_liquidity_amount.ownership >= ownership_to_remove,
-                Error::<T, I>::MoreLiquidity
-            );
-
-            let base_amount = (ownership_to_remove
-                .saturating_mul(pool_liquidity_amount.base_amount))
-            .checked_div(pool_liquidity_amount.ownership)
-            .expect("cannot overflow with positive divisor; qed");
-
-            let quote_amount = (ownership_to_remove
-                .saturating_mul(pool_liquidity_amount.quote_amount))
-            .checked_div(pool_liquidity_amount.ownership)
-            .expect("cannot overflow with positive divisor; qed");
-
             Pools::<T, I>::try_mutate(
                 &base_asset,
                 &quote_asset,
                 |pool_liquidity_amount| -> DispatchResult {
-                    pool_liquidity_amount.base_amount = pool_liquidity_amount
+                    let mut liquidity_amount = pool_liquidity_amount
+                        .take()
+                        .ok_or(Error::<T, I>::PoolDoesNotExist)?;
+                    ensure!(
+                        liquidity_amount.ownership >= ownership_to_remove,
+                        Error::<T, I>::MoreLiquidity
+                    );
+
+                    let base_amount = (ownership_to_remove
+                        .saturating_mul(liquidity_amount.base_amount))
+                    .checked_div(liquidity_amount.ownership)
+                    .expect("cannot overflow with positive divisor; qed");
```

Comment:
shall this expect be replaced and just revert storage changes?

Reply:
Could you please share any example?

Reply:
done